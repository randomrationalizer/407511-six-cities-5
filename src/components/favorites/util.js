export const getCitiesFromOffers = (offers) => {
  const cities = offers.slice().map((offer) => offer.city.name).sort();
  return [...new Set(cities)];
};
