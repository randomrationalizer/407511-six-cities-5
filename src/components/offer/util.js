import dayjs from "dayjs";
import {OfferType} from "../../const";


export const capitalize = (str) => {
  return str.charAt(0).toUpperCase() + str.slice(1);
};

// Возвращает дату в виде строки в формате: 2019-04-24
export const formatDate = (date) => {
  return dayjs(date).format(`YYYY-MM-DD`);
};

// Возвращает дату в виде строки в формате: January 2020
export const humanizeDate = (date) => {
  return dayjs(date).format(`MMMM YYYY`);
};

// Сравнение для сортировки отзывов в хронологическом порядке
export const sortByDate = (a, b) => {
  return dayjs(a.date) - dayjs(b.date);
};

export const getDescriptionSentences = (text) => {
  let description = text.split(`. `);
  description[description.length - 1] = description[description.length - 1].replace(`.`, ``);

  return description;
};

export const isFavoritesCard = (type) => {
  return type === OfferType.FAVORITES;
};

export const isMainPageCard = (type) => {
  return type === OfferType.MAIN;
};
