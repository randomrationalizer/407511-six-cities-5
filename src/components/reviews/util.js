const REVIEW_MIN_LENGTH = 50;
const REVIEW_MAX_LENGTH = 300;
const RATING_MIN = 1;
const RATING_MAX = 5;

export const checkCommentValidity = (comment) => {
  return comment.length >= REVIEW_MIN_LENGTH && comment.length <= REVIEW_MAX_LENGTH ? true : false;
};

export const checkRatingValidity = (rating) => {
  return rating >= RATING_MIN && rating <= RATING_MAX ? true : false;
};
