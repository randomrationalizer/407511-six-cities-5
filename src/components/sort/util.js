import {SortType} from "../../const";

export const sortTypeToTitle = {
  [SortType.DEFAULT]: `Popular`,
  [SortType.PRICE_UP]: `Price: low to high`,
  [SortType.PRICE_DOWN]: `Price: high to low`,
  [SortType.RATING_DOWN]: `Top rated first`
};
