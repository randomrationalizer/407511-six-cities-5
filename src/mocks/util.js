import configureStore from "redux-mock-store";

export const noop = () => {};

export const getMockStore = (store) => {
  const storeConfig = configureStore();
  return storeConfig(store);
};

export const nope = () => {
  return Promise.resolve({});
};
