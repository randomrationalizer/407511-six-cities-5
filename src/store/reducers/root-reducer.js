import {combineReducers} from "redux";
import {appData} from "./app-data/app-data";
import {user} from "./user/user";

const NameSpace = {
  DATA: `DATA`,
  USER: `USER`
};

export default combineReducers({
  [NameSpace.DATA]: appData,
  [NameSpace.USER]: user
});
