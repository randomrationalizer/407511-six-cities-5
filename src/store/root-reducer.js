import {combineReducers} from "redux";
import {appData} from "./app-data/app-data";
import {user} from "./user/user";
import {currentOffer} from "./current-offer/current-offer";
import {loadStatus} from "./load-status/load-status";

export const NameSpace = {
  APP_DATA: `APP_DATA`,
  USER: `USER`,
  CURRENT_OFFER: `CURRENT_OFFER`,
  LOAD_STATUS: `LOAD_STATUS`
};

export const rootReducer = combineReducers({
  [NameSpace.APP_DATA]: appData,
  [NameSpace.USER]: user,
  [NameSpace.CURRENT_OFFER]: currentOffer,
  [NameSpace.LOAD_STATUS]: loadStatus
});
