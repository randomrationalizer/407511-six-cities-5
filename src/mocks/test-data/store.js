import configureStore from "redux-mock-store";
import {NameSpace} from "../../store/root-reducer";
import mockOffers from "./offers";
import mockReviews from "./reviews";
import mockNearbyOffers from "./nearby-offers";
import mockCurrentOffer from "./current-offer";
import mockUserInfo from "./user-info";
import mockFavorites from "./favorites";
import mockCities from "./cities";
import {SortType, AuthorizationStatus, DEFAULT_CITY} from "../../const";


export default {
  [NameSpace.APP_DATA]: {
    currentCity: DEFAULT_CITY,
    allOffers: mockOffers,
    cities: mockCities,
    currentSort: SortType.DEFAULT
  },
  [NameSpace.USER]: {
    authorizationStatus: AuthorizationStatus.AUTH,
    isAuthRequestComplete: true,
    userInfo: mockUserInfo,
    favorites: mockFavorites
  },
  [NameSpace.CURRENT_OFFER]: {
    offer: mockCurrentOffer,
    reviews: mockReviews,
    nearbyOffers: mockNearbyOffers
  },
  [NameSpace.LOAD_STATUS]: {
    isOffersLoaded: true,
    isFavoritesLoaded: true,
    isCurrentOfferLoaded: true,
    errorMessage: null,
    isLoadFinished: true
  }
};
