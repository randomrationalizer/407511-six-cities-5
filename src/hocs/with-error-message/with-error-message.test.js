import React from "react";
import {Provider} from "react-redux";
import PropTypes from "prop-types";
import renderer from "react-test-renderer";
import withErrorMessage, {withErrorMessage as withErrorMessageWithoutStore} from "./with-error-message";
import store from "../../mocks/test-data/store";
import {getMockStore, noop} from "../../mocks/util";
import {NameSpace} from "../../store/reducers/root-reducer";
import {extend} from "../../utils/common";


const message = `Error`;
const mockStore = getMockStore(store);

const MockComponent = (props) => {
  const {children} = props;

  return (
    <div>
      {children}
    </div>
  );
};

MockComponent.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};

const MockComponentWrapped = withErrorMessage(MockComponent);
const MockComponentWrappedWithoutStore = withErrorMessageWithoutStore(MockComponent);

describe(`Should withErrorMessage wrapped component renders correctly`, () => {
  it(`with error`, () => {
    const tree = renderer
      .create(
          <MockComponentWrappedWithoutStore
            message={message}
            closeErrorModal={noop}
          >
            <React.Fragment />
          </MockComponentWrappedWithoutStore>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`without error`, () => {
    const tree = renderer
      .create(
          <MockComponentWrappedWithoutStore
            message={null}
            closeErrorModal={noop}
          >
            <React.Fragment />
          </MockComponentWrappedWithoutStore>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe(`Should withErrorMessage wrapped component connected to store renders correctly`, () => {
  it(`without error`, () => {
    const tree = renderer
      .create(
          <Provider store={mockStore}>
            <MockComponentWrapped>
              <React.Fragment />
            </MockComponentWrapped>
          </Provider>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`with error`, () => {
    const updatedStore = extend(store, {
      [NameSpace.LOAD_STATUS]: extend(store[NameSpace.LOAD_STATUS], {
        errorMessage: message
      })
    });
    const updatedMockStore = getMockStore(updatedStore);

    const tree = renderer
      .create(
          <Provider store={updatedMockStore}>
            <MockComponentWrapped>
              <React.Fragment />
            </MockComponentWrapped>
          </Provider>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
