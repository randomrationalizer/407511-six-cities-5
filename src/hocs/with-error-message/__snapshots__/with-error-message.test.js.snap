// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should withErrorMessage wrapped component connected to store renders correctly with error 1`] = `
Array [
  <div
    className="error"
  >
    <h2>
      Something went wrong:
    </h2>
    <p>
      Error
    </p>
    <button
      className="error-close"
      onClick={[Function]}
      type="button"
    >
      <span
        className="visually-hidden"
      >
        Close
      </span>
    </button>
  </div>,
  <div />,
]
`;

exports[`Should withErrorMessage wrapped component connected to store renders correctly without error 1`] = `<div />`;

exports[`Should withErrorMessage wrapped component renders correctly with error 1`] = `
Array [
  <div
    className="error"
  >
    <h2>
      Something went wrong:
    </h2>
    <p>
      Error
    </p>
    <button
      className="error-close"
      onClick={[Function]}
      type="button"
    >
      <span
        className="visually-hidden"
      >
        Close
      </span>
    </button>
  </div>,
  <div />,
]
`;

exports[`Should withErrorMessage wrapped component renders correctly without error 1`] = `<div />`;
